pydantic/main.py

```diff
@@ -289,7 +290,19 @@ def is_untouched(v: Any) -> bool:
         cls = super().__new__(mcs, name, bases, new_namespace, **kwargs)
         # set __signature__ attr only for model class, but not for its instances
         cls.__signature__ = ClassAttribute('__signature__', generate_model_signature(cls.__init__, fields, config))
-        cls.__try_update_forward_refs__()
+
+        localns = {}
+
+        # Pull in the currently defined local namespace and any namespaces above 
+        if hasattr(sys, "_getframe"):
+            frame = sys._getframe(1)
+            while frame is not None:
+                for k, v in frame.f_locals.items():
+                    if k not in localns:
+                        localns[k] = v
+                frame = frame.f_back
+
```

Comment:
This is the "all enclosing scopes version" - if you are only worried about adding local definitions, at least as first, this simpler version would suffice:

```suggestion

localns = sys._getframe(1).f_locals if hasattr(sys, "_getframe") else {}
```

Also, this doesn't handle classes defined directly within classes, which might have been important too? That could be added as well by including the class namespace (`new_namespace` above).

Reply:
`frame.f_back` goes to the calling function. This isn't 'enclosing' scopes, it's all scopes being called on the stack, which is very likely to cause problems as it may use completely unrelated names.

I gave a suggestion for one way to choose which frame to get locals from here: https://github.com/samuelcolvin/pydantic/issues/2678#issuecomment-822685329

> Also, this doesn't handle classes defined directly within classes

I'm not 100% sure what you have in mind, but inner classes don't have direct access to the enclosing class scope:

```python
class A:
    x = 1
    class B:
        print(x)  # NameError
```

Reply:
Yes, you are right, going "up" the stack would be too much. Getting the immediate locals, though, would help quite a bit with most common issues.

I was assuming the issue with typing in classes looked more like this:

```python
class A:
    class B: pass

    x: B
```

And you want to resolve `B` from `A`.